// react-pathjs-chart/lib/charts/Line.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _palleteColorsJs = require('../pallete/Colors.js');

var _palleteColorsJs2 = _interopRequireDefault(_palleteColorsJs);

var _componentOptionsJs = require('../component/Options.js');

var _componentOptionsJs2 = _interopRequireDefault(_componentOptionsJs);

var _fontAdapterJs = require('../fontAdapter.js');

var _fontAdapterJs2 = _interopRequireDefault(_fontAdapterJs);

var Axis = require('../component/Axis');
var Path = require('paths-js/path');

function cyclic(coll, i) {
    return coll[i % coll.length];
}

var LineChart = (function (_React$Component) {
    _inherits(LineChart, _React$Component);

    function LineChart(props, chartType) {
        _classCallCheck(this, LineChart);

        _get(Object.getPrototypeOf(LineChart.prototype), 'constructor', this).call(this, props);
        this.chartType = chartType;
        this.state = { finished: true };
    }

    _createClass(LineChart, [{
        key: 'getMaxAndMin',
        value: function getMaxAndMin(chart, key, scale) {
            var maxValue;
            var minValue;
            _lodash2['default'].each(chart.curves, function (serie) {
                var values = _lodash2['default'].map(serie.item, function (item) {
                    return item[key];
                });

                var max = _lodash2['default'].max(values);
                if (maxValue === undefined || max > maxValue) maxValue = max;
                var min = _lodash2['default'].min(values);
                if (minValue === undefined || min < minValue) minValue = min;
            });
            return {
                minValue: minValue,
                maxValue: maxValue,
                min: scale(minValue),
                max: scale(maxValue)
            };
        }
    }, {
        key: 'color',
        value: function color(i) {
            var color = this.props.options.color;
            if (!_lodash2['default'].isString(this.props.options.color)) color = color.color;
            var pallete = this.props.pallete || _palleteColorsJs2['default'].mix(color || '#9ac7f7');
            return _palleteColorsJs2['default'].string(cyclic(pallete, i));
        }
    }, {
        key: 'render',
        value: function render() {
            var noDataMsg = this.props.noDataMessage || "No data available";
            if (this.props.data === undefined) return _react2['default'].createElement(
                'span',
                null,
                noDataMsg
            );

            var options = new _componentOptionsJs2['default'](this.props);

            var accessor = function accessor(key) {
                return function (x) {
                    return x[key];
                };
            };
            var chart = this.chartType({
                data: this.props.data,
                xaccessor: accessor(this.props.xKey),
                yaccessor: accessor(this.props.yKey),
                width: options.chartWidth,
                height: options.chartHeight,
                closed: false
            });

            var chartArea = {
                x: this.getMaxAndMin(chart, this.props.xKey, chart.xscale),
                y: this.getMaxAndMin(chart, this.props.yKey, chart.yscale),
                margin: options.margin
            };

            var transparent = { opacity: 0.5 };

            var lines = _lodash2['default'].map(chart.curves, (function (c, i) {
                return _react2['default'].createElement('path', { key: 'lines' + i, d: c.line.path.print(), stroke: this.color(i), fill: 'none' });
            }).bind(this));
            var areas = _lodash2['default'].map(chart.curves, (function (c, i) {
                //var transparent = { opacity: 0.5 };
                return _react2['default'].createElement('path', { key: 'areas' + i, d: c.area.path.print(), style: transparent, stroke: 'none', fill: this.color(i) });
            }).bind(this));

            return _react2['default'].createElement(
                'svg',
                { ref: 'vivus', width: options.width, height: options.height },
                _react2['default'].createElement(
                    'g',
                    { transform: "translate(" + options.margin.left + "," + options.margin.top + ")" },
                    this.state.finished ? areas : null,
                    lines,
                    _react2['default'].createElement(Axis, { key: 'x', scale: chart.xscale, options: options.axisX, chartArea: chartArea }),
                    _react2['default'].createElement(Axis, { key: 'y', scale: chart.yscale, options: options.axisY, chartArea: chartArea })
                )
            );
        }
    }]);

    return LineChart;
})(_react2['default'].Component);

exports['default'] = LineChart;
module.exports = exports['default'];
