// react-pathjs-chart/lib/charts/TreeVivus.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _vivus = require('vivus');

var _vivus2 = _interopRequireDefault(_vivus);

var _TreeJs = require('./Tree.js');

var _TreeJs2 = _interopRequireDefault(_TreeJs);

var TreeVivusChart = (function (_TreeChart) {
    _inherits(TreeVivusChart, _TreeChart);

    function TreeVivusChart(props) {
        _classCallCheck(this, TreeVivusChart);

        _get(Object.getPrototypeOf(TreeVivusChart.prototype), 'constructor', this).call(this, props);
        this.state = { finished: false };
    }

    _createClass(TreeVivusChart, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            if (nextProps.replay !== this.props.replay) this.setState({ finished: false });
        }
    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.run();
        }
    }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate(prevProps, prevState) {
            if (!this.state.finished) this.run();
        }
    }, {
        key: 'run',
        value: function run() {
            if (this.refs.vivus === undefined) return;
            var animate = this.props.options && this.props.options.animate || {};
            new _vivus2['default'](_reactDom2['default'].findDOMNode(this.refs.vivus), {
                type: animate.type || 'delayed',
                duration: animate.duration || 'delayed',
                start: 'autostart',
                selfDestroy: true
            }, this.finish.bind(this));
        }
    }, {
        key: 'finish',
        value: function finish() {
            this.setState({ finished: true });
        }
    }]);

    return TreeVivusChart;
})(_TreeJs2['default']);

exports['default'] = TreeVivusChart;
module.exports = exports['default'];
