// redux-promise-wait/create-wait.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _actions = require('./actions');

var warning = function warning(condition, message) {
  if (process.env.NODE_ENV !== 'production' && condition) {
    /* eslint-disable no-console */
    console.warn('[redux-wait] ' + message);
    /* eslint-enable no-console */
  }
};

var mapAction = function mapAction(_ref) {
  var promise = _ref.promise;
  var action = _ref.action;
  return promise.then(function (result) {
    return { action: action, result: result };
  }, function (error) {
    return { action: action, error: error };
  });
};

/**
 * Create a function that awaits all promises in the `redux-wait` store.
 *
 * @function
 *
 * @param {Function} callback A callback function.
 *
 * @param {Object} store A redux store enhanced with `redux-wait`.
 *
 * @param {Number} options.maxIterations The maximum number callback iterations.
 *
 * @param {Number} options.storeName The name of the property containing the
 * `waitStore` on the base store object, this should match the same config
 * option provided to `waitEnhancer()`.
 *
 * @returns {Function} A function that returns a promise.
 */

exports.default = function (callback, store) {
  var _ref2 = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

  var _ref2$maxIterations = _ref2.maxIterations;
  var maxIterations = _ref2$maxIterations === undefined ? 2 : _ref2$maxIterations;
  var _ref2$storeName = _ref2.storeName;
  var storeName = _ref2$storeName === undefined ? 'waitStore' : _ref2$storeName;

  warning(maxIterations < 2, 'A `maxIterations` value of less than 2 will not wait for any actions to ' + 'resolve. Specify a higher value.');

  var waitStore = store[storeName];

  warning(!waitStore, 'The provided store must be created using the `waitEnhancer`.');

  return function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var iterate = function iterate() {
      var count = arguments.length <= 0 || arguments[0] === undefined ? 1 : arguments[0];

      var start = Date.now();

      var renderResult = callback.apply(undefined, args);

      var _waitStore$getState = waitStore.getState();

      var actions = _waitStore$getState.actions;

      waitStore.dispatch((0, _actions.clearActions)());

      if (count < maxIterations) {
        if (actions.length) {
          return Promise.all(actions.map(mapAction)).then(function (results) {
            var duration = Date.now() - start;
            waitStore.dispatch((0, _actions.stats)({
              results: results,
              duration: duration
            }));
            return iterate(count + 1);
          });
        }

        return Promise.resolve(renderResult);
      }

      warning(actions.length, 'Callback completed with unresolved actions. Specify a higher ' + 'value for the `maxIterations` option or reduce the depth of async ' + 'action creator calls.');

      return Promise.resolve(renderResult);
    };

    return iterate();
  };
};
